contracts/utils/structs/DoubleEndedQueue.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "../math/SafeCast.sol";
+
+library DoubleEndedQueue {
+    error Empty();
+    error OutOfBounds();
+
+    struct Bytes32Deque {
+        int128 begin; // inclusive: the first item is at data[begin]
+        int128 end; // exclusive: the last item is at data[end-1]
+        mapping(int128 => bytes32) data;
+    }
+
+    function pushBack(Bytes32Deque storage deque, bytes32 value) internal {
+        int128 backIndex = deque.end;
+        deque.data[backIndex] = value;
+        unchecked {
+            deque.end = backIndex + 1;
+        }
```

Comment:
IMO, the previous implementation
```
unchecked {
     vector.data[vector.end++] = value;
}
```
was way more easy to read/understand (same argument apply to all push and pop functions)

Reply:
I definitely think we want the `+1` unchecked.

Putting everything in the unchecked block might make it cleaner to read (all the logic is indented at the same level)